lua/lazyvim/plugins/extras/lang/python.lua

```diff
@@ -0,0 +1,110 @@
+return {
+  {
+    "nvim-treesitter/nvim-treesitter",
+    opts = function(_, opts)
+      if type(opts.ensure_installed) == "table" then
+        vim.list_extend(opts.ensure_installed, { "ninja", "python", "rst", "toml" })
+      end
+    end,
+  },
+  {
+    "jose-elias-alvarez/null-ls.nvim",
+    opts = function(_, opts)
+      local nls = require("null-ls")
+      opts.sources = opts.sources or {}
+      vim.list_extend(opts.sources, {
+        -- Order of formatters matters. They are used in order of appearance.
+        nls.builtins.formatting.ruff.with({ filetypes = { "python" }, prefer_local = ".venv/bin" }),
+        nls.builtins.formatting.black.with({ filetypes = { "python" }, prefer_local = ".venv/bin" }),
+      })
+    end,
+    dependencies = {
+      "mason.nvim",
+      opts = function(_, opts)
+        opts.ensure_installed = opts.ensure_installed or {}
+        vim.list_extend(opts.ensure_installed, { "black", "ruff" })
```

Comment:
I don't personally load `black` from Mason unless it isn't found on `$PATH` (from my activated `$VIRTUAL_ENV`). 

```lua
formatting.black.with({
  command = function()
    -- if executable is on $PATH, use it
    if vim.fn.executable("black") == 1 then
      return "black"
    end
    -- otherwise, use the executable from Mason
    return mason_registry.get_package("black"):get_install_path()
  end,
}),
```

And I ensure `ruff-lsp` is installed via Mason rather than ensuring that `ruff` is installed. You're also not including `mypy` which is very common, so you're here providing "sane defaults" that might not apply to everyone.

I'm starting to think that maybe as a first step, you should omit ruff and black entirely from this PR and call it a day. It'll be really hard to cater to everyone's need here. I would even go as far as putting in a comment about NOT adding in linters/formatters here.

Reply:
That's why I used the `prefer_local` option. But for now I removed the `null-ls` package entirely from this extra. There is to many differences between everyones configurations.